kik_unofficial/datatypes/xmpp/base_elements.py

```diff
@@ -1,29 +1,303 @@
+import base64
+import binascii
+import uuid
+from typing import Union, final
+
 from bs4 import BeautifulSoup
+from lxml import etree
+from lxml.etree import Element
 
+from kik_unofficial.utilities import jid_utilities
 from kik_unofficial.utilities.cryptographic_utilities import CryptographicUtils
+from kik_unofficial.utilities.kik_server_clock import KikServerClock
 
 
 class XMPPElement:
     def __init__(self):
         self.message_id = CryptographicUtils.make_kik_uuid()
-        self.content_id = CryptographicUtils.make_kik_uuid() #Creating a seprate uuid for content as kik does the same.
 
-    def serialize(self) -> bytes:
+    def serialize(self) -> Union[bytes, Element]:
+        raise NotImplementedError
+
+
+class XMPPOutgoingMessageElement(XMPPElement):
+    def __init__(self, peer_jid: str):
+        super().__init__()
+        self.peer_jid = peer_jid
+        self.is_group = jid_utilities.is_group_jid(peer_jid)
+        self.timestamp = str(KikServerClock.get_server_time())
+        self.message_type = "groupchat" if self.is_group else "chat"
+
+    @final
+    def serialize(self) -> Element:
+        message = etree.Element('message')
+        message.set('type', self.message_type)
+        if self.is_group:
+            message.set('xmlns', 'kik:groups')
+        message.set('to', self.peer_jid)
+        message.set('id', self.message_id)
+        if not self.message_type == 'is-typing':
+            message.set('cts', self.timestamp)
+        self.serialize_message(message)
+        return message
+
+    def serialize_message(self, message: Element) -> None:
+        raise NotImplementedError
+
+    @final
+    def add_kik_element(self, message: Element, push: bool = True, qos: bool = True) -> None:
+        kik = etree.SubElement(message, 'kik')
+        kik.set('push', 'true' if push else 'false')
+        kik.set('qos', 'true' if qos else 'false')
+        kik.set('timestamp', self.timestamp)
+
+    @final
+    def add_request_element(self, message: Element, request_delivered: bool = True, request_read: bool = False) -> None:
+        if not request_read and not request_delivered:
+            return
+        request = etree.SubElement(message, 'request')
+        request.set('xmlns', 'kik:message:receipt')
+        request.set('d', 'true' if request_delivered else 'false')
+        request.set('r', 'true' if request_read else 'false')
+
+    @final
+    def add_empty_element(self, message: Element, name: str) -> None:
+        etree.SubElement(message, name).text = ''
+
+
+class XMPPOutgoingContentMessageElement(XMPPOutgoingMessageElement):
```

Comment:
Maybe add a docstring explaining what is a content message element

Reply:
Resolved in c15df46c426a39b08273efec4903c79faa75842c